Painted images don't get tone mapped down to the level supported by the display.

mattwoodrow opened the pull request:
#### 75fb0f2929ea8df1f8984abb911e172a16eca764
<pre>
Painted images don&apos;t get tone mapped down to the level supported by the display.
<a href="https://bugs.webkit.org/show_bug.cgi?id=294419">https://bugs.webkit.org/show_bug.cgi?id=294419</a>
&lt;<a href="https://rdar.apple.com/152695267">rdar://152695267</a>&gt;

Reviewed by Mike Wyrzykowski.

Add support for getting the currently available EDR headroom to PlatformScreen,
and synchronize it to all processes. Use
NSApplicationDidChangeScreenParametersNotification to detect change on macos,
use a poll on every displaylink update on iOS.

When painting via RemoteDisplayListRecorderProxy, clamp image headroom to the
maximum currently available on the display, and keep track of what the highest
painted value was (and if it was clamped). Theses values get stored on the
RemoteLayerBackingStore.

When display headroom changes, trigger a compositing update for all layers. Adds
a new function &apos;setNeedsDisplayIfHeadroomExceeds&apos; to GraphicsLayer (and layers
below it) that conditionally mark repaint if the new display headroom would
apply different clamping to what is currently painted into the layer.

Some of this is a bit awkward since we don&apos;t really know what is painted into a
layer in advance, so using the recorded to track what HDR painting occurred and
keep it available for the next paint.

This approach lets us track headroom per-tile, and only repaint affected tiles
when the display headroom. The downside is that a content change to remove an
HDR image won&apos;t drop the max headroom for tiles that it was previouly drawn to,
so those will get repainted again if the display headroom changes (once, and
then the cached headroom will be correct).

We also serialize the resulting headroom of the layer to the UI process via
RemoteLayerBackingStoreProperties. We use CAIOSurface to cache a CoreAnimation
wrapper around an IOSurface, and we need to recreate these if the IOSurface
headroom metadata has changed.

* Source/WebCore/PAL/pal/cocoa/QuartzCoreSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/QuartzCoreSoftLink.mm:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::currentFrameHeadroom const):
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/page/Page.cpp:
(WebCore::m_presentingApplicationBundleIdentifier):
(WebCore::Page::screenPropertiesDidChange):
(WebCore::Page::updateDisplayEDRHeadroom):
* Source/WebCore/page/Page.h:
(WebCore::Page::displayEDRHeadroom const):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::setMaxEDRHeadroom):
(WebCore::GraphicsContext::maxPaintedHeadroom const):
(WebCore::GraphicsContext::paintingClampedHeadroom const):
(WebCore::GraphicsContext::clearMaxPaintedHeadroom):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::currentFrameHeadroom const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
(WebCore::GraphicsContextCG::setMaxEDRHeadroom):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::asCAIOSurfaceLayerContents const):
(WebCore::IOSurface::setContentEDRHeadroom):
(WebCore::IOSurface::contentEDRHeadroom const):
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::currentEDRHeadroomForDisplay):
(WebCore::maxEDRHeadroomForDisplay):
(WebCore::collectScreenProperties):
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::collectScreenProperties):
(WebCore::currentEDRHeadroomForDisplay):
(WebCore::maxEDRHeadroomForDisplay):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::RenderElement):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::hasHDRImages const):
(WebCore::RenderElement::setHasHDRImages):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::paintContents):
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::dump const):
(WebKit::RemoteLayerBackingStore::setNeedsDisplay):
(WebKit::RemoteLayerBackingStore::setNeedsDisplayIfHeadroomExceeds):
(WebKit::RemoteLayerBackingStore::supportsPartialRepaint const):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::updateCachedBuffers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::dump const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
(WebKit::WebProcessPool::willRefreshDisplay):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::didRefreshDisplay):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::willRefreshDisplay):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::drawNativeImageInternal):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
(WebKit::RemoteDisplayListRecorderProxy::paintingClampedHeadroom const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setNeedsDisplayIfHeadroomExceeds):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.cpp:
(WebKit::PlatformCALayerRemoteTiledBacking::setNeedsDisplayIfHeadroomExceeds):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.h:

Canonical link: <a href="https://commits.webkit.org/296360@main">https://commits.webkit.org/296360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c442922882acac16f8677af6c70278fdea401fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108296 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18379 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113505 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110259 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28646 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82223 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111244 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22703 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97549 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62659 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22116 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58237 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100859 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92071 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116627 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106855 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35350 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26029 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91255 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93826 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91056 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23202 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35937 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13708 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35252 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40790 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131141 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34969 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35595 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38323 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36650 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5753,4 +5760,26 @@ bool Page::requiresUserGestureForVideoPlayback() const
+            continue;
+
+        view->setDescendantsNeedUpdateBackingAndHierarchyTraversal();
+
```
Blank line.

smfr replied:
I don't think this is hitting all frames, and should.

smfr commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -239,6 +239,13 @@ class GraphicsContext {
+#if HAVE(SUPPORT_HDR_DISPLAY)
+    virtual void setMaxEDRHeadroom(std::optional<float>) { }
+    virtual float maxPaintedHeadroom() const { return 1; }
+    virtual bool paintingClampedHeadroom() const { return false; }
```
I think this should be `hasPaintedClampedHeadroom()`.

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -743,6 +743,12 @@ void GraphicsLayerCA::setDrawsHDRContent(bool drawsHDRContent)
+
+void GraphicsLayerCA::setNeedsDisplayIfHeadroomExceeds(float headroom)
+{
+    if (protectedLayer()->setNeedsDisplayIfHeadroomExceeds(headroom))
```
This is confusing; why does `GraphicsLayerCA` need to call `setNeedsDisplay()` if `setNeedsDisplayIfHeadroomExceeds()` already did it? Maybe `setNeedsDisplayIfHeadroomExceeds()` should just be `needsRepaintIfHeadroomExceeds()`?

mattwoodrow replied:
Forgot to get back to fixing that!

I think it doesn't, and doing so was actively bad, since it'd mean we repaint tiles that don't have HDR content.

We do however need the feedback that something changed, so that we can schedule a flush (rendering update), in case this happened outside of one.

smfr commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -410,8 +410,12 @@ static IntSize computeMaximumSurfaceSize()
+    if (PAL::canLoad_QuartzCore_CAIOSurfaceCreate()) {
+        auto result = adoptCF(CAIOSurfaceCreate(m_surface.get()));
+        if (PAL::canLoad_QuartzCore_CAIOSurfaceReloadColorAttributes())
+            CAIOSurfaceReloadColorAttributes(result.get());
```
Maybe comment to say what this does.

smfr commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -685,6 +689,28 @@ static IntSize computeMaximumSurfaceSize()
+{
+    if (!m_contentEDRHeadroom) {
+        m_contentEDRHeadroom = 1;
+        if (auto valueNumber = dynamic_cf_cast<CFNumberRef>(IOSurfaceCopyValue(m_surface.get(), CFSTR("IOSurfaceContentHeadroom"))))
```
I think you need an `adoptCF` in here since presumably `IOSurfaceCopyValue` retains.

smfr replied:
If the bot doesn't catch this, let Ryosuke know.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -623,8 +661,11 @@ static bool hasValue(const ImageBufferBackendHandle& backendHandle)
-        m_contentsBuffer = layerContentsBufferFromBackendHandle(WTFMove(*m_bufferHandle), LayerContentsType::CachedIOSurface);
-        cachedBuffers.append({ *m_frontBufferInfo, m_contentsBuffer });
+        if (auto surface = WebCore::IOSurface::createFromSendRight(std::get<MachSendRight>(*std::exchange(m_bufferHandle, std::nullopt)))) {
+            surface->setContentEDRHeadroom(m_edrHeadroom);
```
Need `#if HAVE(SUPPORT_HDR_DISPLAY)`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
```diff
@@ -229,6 +229,9 @@ - (void)didChangeNominalFramesPerSecond
 void RemoteLayerTreeDrawingAreaProxyIOS::didRefreshDisplay()
 {
+    if (RefPtr page = this->page())
+        page->willRefreshDisplay();
```
Confusing for a "did" to call a "will".

weinig commented on Source/WebCore/page/Page.cpp:
```diff
@@ -507,6 +507,10 @@ Page::Page(PageConfiguration&& pageConfiguration)
 #endif
 
+#if HAVE(SUPPORT_HDR_DISPLAY)
+    updateDisplayEDRHeadroom();
```
In some places you are using the un-adorned term "Headroom" and in others you are using "EDRHeadroom". Are these distinct concepts?

mwyrzykowski commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -685,6 +692,30 @@ static IntSize computeMaximumSurfaceSize()
+
+    LOG_WITH_STREAM(HDR, stream << "IOSurface::setContentEDRHeadroom " << this << " " << headroom);
+    m_contentEDRHeadroom = headroom;
+    IOSurfaceSetValue(m_surface.get(), CFSTR("IOSurfaceContentHeadroom"), @(headroom));
```
should we define `"IOSurfaceContentHeadroom"` in `IOSurfaceSPI.h` like the other IOSurface properties we use?

mattwoodrow replied:
Indeed we should, done!

mwyrzykowski commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -685,6 +692,30 @@ static IntSize computeMaximumSurfaceSize()
+    return m_contentEDRHeadroom;
+}
+
+void IOSurface::loadContentEDRHeadroom()
```
is this member function used anywhere? I could not locate a call site

mattwoodrow replied:
It is not. I just had it written (for an earlier version of the PR) and left it in.

mattwoodrow committed:
Painted images don't get tone mapped down to the level supported by the display.
https://bugs.webkit.org/show_bug.cgi?id=294419
<rdar://152695267>

Reviewed by Mike Wyrzykowski.

Add support for getting the currently available EDR headroom to PlatformScreen,
and synchronize it to all processes. Use
NSApplicationDidChangeScreenParametersNotification to detect change on macos,
use a poll on every displaylink update on iOS.

When painting via RemoteDisplayListRecorderProxy, clamp image headroom to the
maximum currently available on the display, and keep track of what the highest
painted value was (and if it was clamped). Theses values get stored on the
RemoteLayerBackingStore.

When display headroom changes, trigger a compositing update for all layers. Adds
a new function 'setNeedsDisplayIfHeadroomExceeds' to GraphicsLayer (and layers
below it) that conditionally mark repaint if the new display headroom would
apply different clamping to what is currently painted into the layer.

Some of this is a bit awkward since we don't really know what is painted into a
layer in advance, so using the recorded to track what HDR painting occurred and
keep it available for the next paint.

This approach lets us track headroom per-tile, and only repaint affected tiles
when the display headroom. The downside is that a content change to remove an
HDR image won't drop the max headroom for tiles that it was previouly drawn to,
so those will get repainted again if the display headroom changes (once, and
then the cached headroom will be correct).

We also serialize the resulting headroom of the layer to the UI process via
RemoteLayerBackingStoreProperties. We use CAIOSurface to cache a CoreAnimation
wrapper around an IOSurface, and we need to recreate these if the IOSurface
headroom metadata has changed.

* Source/WebCore/PAL/pal/cocoa/QuartzCoreSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/QuartzCoreSoftLink.mm:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::currentFrameHeadroom const):
* Source/WebCore/loader/cache/CachedImage.h:
* Source/WebCore/page/Page.cpp:
(WebCore::m_presentingApplicationBundleIdentifier):
(WebCore::Page::screenPropertiesDidChange):
(WebCore::Page::updateDisplayEDRHeadroom):
* Source/WebCore/page/Page.h:
(WebCore::Page::displayEDRHeadroom const):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::setMaxEDRHeadroom):
(WebCore::GraphicsContext::maxPaintedHeadroom const):
(WebCore::GraphicsContext::paintingClampedHeadroom const):
(WebCore::GraphicsContext::clearMaxPaintedHeadroom):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::currentFrameHeadroom const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::setNeedsDisplayIfHeadroomExceeds):
* Source/WebCore/platform/graphics/ca/TileGrid.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
(WebCore::GraphicsContextCG::setMaxEDRHeadroom):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::asCAIOSurfaceLayerContents const):
(WebCore::IOSurface::setContentEDRHeadroom):
(WebCore::IOSurface::contentEDRHeadroom const):
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::currentEDRHeadroomForDisplay):
(WebCore::maxEDRHeadroomForDisplay):
(WebCore::collectScreenProperties):
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::collectScreenProperties):
(WebCore::currentEDRHeadroomForDisplay):
(WebCore::maxEDRHeadroomForDisplay):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::RenderElement):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::hasHDRImages const):
(WebCore::RenderElement::setHasHDRImages):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDrawsContent):
(WebCore::RenderLayerBacking::paintContents):
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::encode const):
(WebKit::RemoteLayerBackingStoreProperties::dump const):
(WebKit::RemoteLayerBackingStore::setNeedsDisplay):
(WebKit::RemoteLayerBackingStore::setNeedsDisplayIfHeadroomExceeds):
(WebKit::RemoteLayerBackingStore::supportsPartialRepaint const):
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::updateCachedBuffers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::dump const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
(WebKit::WebProcessPool::willRefreshDisplay):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeDrawingAreaProxyIOS.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyIOS::didRefreshDisplay):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::willRefreshDisplay):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::drawNativeImageInternal):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
(WebKit::RemoteDisplayListRecorderProxy::paintingClampedHeadroom const):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setNeedsDisplayIfHeadroomExceeds):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.cpp:
(WebKit::PlatformCALayerRemoteTiledBacking::setNeedsDisplayIfHeadroomExceeds):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.h:

Canonical link: https://commits.webkit.org/296360@main
